modules/videoio/src/cap_msmf.cpp

```diff
@@ -997,71 +1239,177 @@ bool CvCapture_MSMF::openFinalize_(const VideoCaptureParameters* params)
+        if (streamIndex != dwVideoStreamIndex)
+            break;
+        if (flags & (MF_SOURCE_READERF_ERROR | MF_SOURCE_READERF_ALLEFFECTSREMOVED | MF_SOURCE_READERF_ENDOFSTREAM))
+            break;
+        if (videoSample)
+            break;
+        if (flags & MF_SOURCE_READERF_STREAMTICK)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream tick detected. Retrying to grab the frame");
+        }
+    }
+    if (SUCCEEDED(hr))
+    {
+        if (streamIndex != dwVideoStreamIndex)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Wrong stream read. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ERROR)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream reading error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ALLEFFECTSREMOVED)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream decoding error. Abort capturing");
+            close();
+        }
+        else if (flags & MF_SOURCE_READERF_ENDOFSTREAM)
+        {
+            CV_LOG_DEBUG(NULL, "videoio(MSMF): End of stream detected");
+        }
+        else
+        {
+            nFrame++;
+            videoSample->GetSampleDuration(&curVideoTime);
+            allTime += curVideoTime;
+            if (duration - allTime < curVideoTime)
+            {
+                lastFrame = true;
+            }
+            if (flags & MF_SOURCE_READERF_NEWSTREAM)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): New stream detected");
+            }
+            if (flags & MF_SOURCE_READERF_NATIVEMEDIATYPECHANGED)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream native media type changed");
+            }
+            if (flags & MF_SOURCE_READERF_CURRENTMEDIATYPECHANGED)
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): Stream current media type changed");
+            }
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CvCapture_MSMF::grabAudioFrame()
+{
+    DWORD streamIndex,  flags;
+    HRESULT hr;
+    bool returnFlag = false;
+    audioSamples.clear();
+    curAudioTime = 0;
+    int numberOfSamples = -1;
+    while (curAudioTime <= curVideoTime)
+    {
+        if (residualTime > curVideoTime)
+            break;
+        audioSamples.push_back(NULL);
+        numberOfSamples++;
+        for (;;)
         {
             CV_TRACE_REGION("ReadSample");
             if (!SUCCEEDED(hr = videoFileSource->ReadSample(
-                dwStreamIndex, // Stream index.
+                dwAudioStreamIndex, // Stream index.
                 0,             // Flags.
                 &streamIndex,  // Receives the actual stream index.
                 &flags,        // Receives status flags.
-                &sampleTime,   // Receives the time stamp.
-                &videoSample   // Receives the sample or NULL.
+                NULL,   // Receives the time stamp.
```

Comment:
> Why audio ignores timestamp?
> 
> This is why the current logic fails on [`sample_960x400_ocean_with_audio.mp4`](https://filesamples.com/samples/video/mp4/sample_960x400_ocean_with_audio.mp4)
> 
> Add debug dump of values from ReadSample:
> 
> ```
> CvCapture_MSMF::initStream Init stream 1 with MediaType (960x400 @ 23.976) MFVideoFormat_RGB32
> CvCapture_MSMF::initStream Init stream 0 with MediaType (0x0 @ 1) ☺
> CvCapture_MSMF::grabVideoFrame ReadSample(video): stream=1 time=1668332
> CvCapture_MSMF::grabVideoFrame ReadSample(video): GetSampleDuration=417083
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=213333
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=206349
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=419682
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=213379
> ...
> CvCapture_MSMF::grabVideoFrame ReadSample(video): stream=1 time=2085415
> CvCapture_MSMF::grabVideoFrame ReadSample(video): GetSampleDuration=417083
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=633061
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=213379
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=846441
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=213152
> ...
> CvCapture_MSMF::grabVideoFrame ReadSample(video): stream=1 time=2502498
> CvCapture_MSMF::grabVideoFrame ReadSample(video): GetSampleDuration=417083
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=1059593
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=213379
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): stream=0 time=1272972
> CvCapture_MSMF::grabAudioFrame ReadSample(audio): GetSampleDuration=213379
> ```
> 
> Audio stream starts 4+ frames before video. We need to capture all of this audio data and return with the first video frame.
Audio ignores timestamp because there was no need to use it. I will fix it.